src/rgw/rgw_data_sync.cc

```diff
@@ -771,6 +765,59 @@ int RGWRemoteDataLog::read_recovering_shards(const DoutPrefixProvider *dpp, cons
   return ret;
 }
 
+namespace RGWRDL {
+class DataSyncInitCR : public RGWCoroutine {
+  RGWDataSyncCtx* const sc;
+  const uint32_t num_shards;
+  uint64_t instance_id;
+  const RGWSyncTraceNodeRef& tn;
+  rgw_data_sync_status* const sync_status;
+
+  boost::intrusive_ptr<RGWContinuousLeaseCR> lease_cr;
+
+public:
+
+  DataSyncInitCR(RGWDataSyncCtx* sc, uint32_t num_shards, uint64_t instance_id,
+		 const RGWSyncTraceNodeRef& tn,
+		 rgw_data_sync_status* sync_status)
+    : RGWCoroutine(sc->cct), sc(sc), num_shards(num_shards),
+      instance_id(instance_id), tn(tn), sync_status(sync_status) {}
+
+  ~DataSyncInitCR() override {
+    if (lease_cr) {
+      lease_cr->abort();
+    }
+  }
+
+  int operate(const DoutPrefixProvider *dpp) override {
+    reenter(this) {
+      lease_cr.reset(
+	RGWInitDataSyncStatusCoroutine::continuous_lease_cr(sc, this));
+
+      yield spawn(lease_cr.get(), false);
+      while (!lease_cr->is_locked()) {
+	if (lease_cr->is_done()) {
+	  tn->log(5, "ERROR: failed to take data sync status lease");
+	  set_status("lease lock failed, early abort");
+	  drain_all();
+	  return set_cr_error(lease_cr->get_ret_status());
+	}
+	tn->log(5, "waiting on data sync status lease");
+	yield set_sleeping(true);
+      }
+      tn->log(5, "acquired data sync status lease");
+      new RGWInitDataSyncStatusCoroutine(sc, num_shards, instance_id, tn,
+					 sync_status, lease_cr, nullptr);
```

Comment:
needs a `yield call()` to actually run it, and a check for its `retcode`

other `RGWInitDataSyncStatusCoroutine` calls are using `RGWObjVersionTracker` to detect racing writes. for that to work, this call also needs to supply one. because we want to overwrite any existing status, we just have to call `objv_tracker.generate_new_write_ver(cct)` on a default-constructed one. that results in a `cls_version_set()` call that causes other racing `cls_version_check()`s to fail with `ECANCELED`

Reply:
Fixed!